pkg/csi/cinder/controllerserver.go

```diff
@@ -392,6 +475,78 @@ func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		klog.Errorf("Failed to WaitSnapshotReady: %v", err)
 		return nil, status.Errorf(codes.Internal, "CreateSnapshot failed with error %v", err)
 	}
+	// If snapshotType is 'backup', create a backup from the snapshot and delete the snapshot.
+	if snapshotType == "backup" {
+
+		// Add cluster ID to the snapshot metadata
+		properties := map[string]string{cinderCSIClusterIDKey: cs.Driver.cluster}
+
+		// see https://github.com/kubernetes-csi/external-snapshotter/pull/375/
+		// Also, we don't want to tag every param but we still want to send the
+		// 'force-create' flag to openstack layer so that we will honor the
+		// force create functions
+		for _, mKey := range []string{"csi.storage.k8s.io/volumesnapshot/name", "csi.storage.k8s.io/volumesnapshot/namespace", "csi.storage.k8s.io/volumesnapshotcontent/name", openstack.SnapshotForceCreate, openstack.SnapshotType, openstack.BackupMaxDurationPerGB} {
+			if v, ok := req.Parameters[mKey]; ok {
+				properties[mKey] = v
+			}
+		}
+
+		// We repeat the check from the snapshot creation in case Cinder-CSI gets restarted while a backup is in progress.
```

Comment:
I'm not sure I understand here. Why do we need to check it again?

Reply:
Basically, the previous check stops the controller from creating a new snapshot if either a snapshot or a backup with the same name exists
If a snapshot with the sane name exists, it skips to `WaitSnapshotReady`.
If a backup with the same name exists it sets `skipSnapshot` to true and skips waiting for the snapshot. It also skips deleting it. (I just added this `skipSnapshot` part)

The logic for the backup starts after we know that the snapshot is ready. We do the same thing where we avoid creating a new backup if a backup with the same name already exists, and skip to `WaitBackupReady`.

We always do these checks because the controller pod might restart after a snapshot or backup is created, but before a `CreateSnapshotResponse` has been returned.